File: src/cjhx_agile_workflow/workflows.py
from __future__ import annotations

import json
import uuid
from dataclasses import asdict, dataclass, field
from pathlib import Path
from typing import Any

from .audit import redact
from .errors import ValidationError
from .models import utc_now
from .skills import SkillRuntime
from .storage import Workspace


@dataclass(frozen=True)
class WorkflowStep:
    id: str
    skill: str
    input: dict[str, Any]


@dataclass(frozen=True)
class WorkflowDefinition:
    id: str
    version: str
    name: str
    steps: tuple[WorkflowStep, ...]

    @classmethod
    def load(cls, path: str | Path) -> WorkflowDefinition:
        source = Path(path)
        try:
            value = json.loads(source.read_text(encoding="utf-8"))
        except FileNotFoundError as error:
            raise ValidationError(f"workflow not found: {source}") from error
        except json.JSONDecodeError as error:
            raise ValidationError(f"invalid workflow JSON: {error}") from error
        if not isinstance(value, dict):
            raise ValidationError("workflow must be a JSON object")
        missing = [key for key in ("id", "version", "name", "steps") if key not in value]
        if missing:
            raise ValidationError(f"missing workflow fields: {', '.join(missing)}")
        if not isinstance(value["steps"], list) or not value["steps"]:
            raise ValidationError("workflow steps must be a non-empty array")
        steps: list[WorkflowStep] = []
        seen: set[str] = set()
        for raw in value["steps"]:
            if not isinstance(raw, dict) or not all(key in raw for key in ("id", "skill", "input")):
                raise ValidationError("each workflow step requires id, skill, and input")
            if raw["id"] in seen:
                raise ValidationError(f"duplicate workflow step id: {raw['id']}")
            if not isinstance(raw["input"], dict):
                raise ValidationError(f"workflow step input must be an object: {raw['id']}")
            seen.add(raw["id"])
            steps.append(WorkflowStep(str(raw["id"]), str(raw["skill"]), raw["input"]))
        return cls(str(value["id"]), str(value["version"]), str(value["name"]), tuple(steps))


@dataclass(frozen=True)
class WorkflowRun:
    id: str
    workflow_id: str
    workflow_version: str
    change_id: str | None
    status: str
    started_at: str
    completed_at: str
    input: dict[str, Any]
    steps: list[dict[str, Any]] = field(default_factory=list)
    output: dict[str, Any] = field(default_factory=dict)
    error: str | None = None

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


class WorkflowRuntime:
    def __init__(self, workspace: Workspace, skills: SkillRuntime) -> None:
        self.workspace = workspace
        self.skills = skills

    def run(
        self,
        definition: WorkflowDefinition,
        payload: dict[str, Any],
        *,
        change_id: str | None = None,
        approved_steps: set[str] | None = None,
    ) -> WorkflowRun:
        run_id = f"workflow-run-{uuid.uuid4().hex}"
        started = utc_now()
        context: dict[str, Any] = {"input": payload, "steps": {}}
        records: list[dict[str, Any]] = []
        approved_steps = approved_steps or set()
        try:
            for step in definition.steps:
                resolved = self._resolve(step.input, context)
                skill_run = self.skills.run(
                    step.skill,
                    resolved,
                    change_id=change_id,
                    approved=step.id in approved_steps,
                )
                context["steps"][step.id] = {"output": skill_run.output, "runId": skill_run.id}
                records.append(
                    {
                        "id": step.id,
                        "skill": step.skill,
                        "skillRunId": skill_run.id,
                        "status": skill_run.status,
                        "output": skill_run.output,
                    }
                )
            run = WorkflowRun(
                id=run_id,
                workflow_id=definition.id,
                workflow_version=definition.version,
                change_id=change_id,
                status="succeeded",
                started_at=started,
                completed_at=utc_now(),
                input=redact(payload),
                steps=redact(records),
                output=redact(context["steps"][definition.steps[-1].id]["output"]),
            )
        except Exception as error:
            run = WorkflowRun(
                id=run_id,
                workflow_id=definition.id,
                workflow_version=definition.version,
                change_id=change_id,
                status="failed",
                started_at=started,
                completed_at=utc_now(),
                input=redact(payload),
                steps=redact(records),
                error=f"{type(error).__name__}: {error}",
            )
            self.workspace.initialize()
            self.workspace.write_json(self.workspace.runs / f"{run.id}.json", run.to_dict())
            raise
        self.workspace.write_json(self.workspace.runs / f"{run.id}.json", run.to_dict())
        return run

    @classmethod
    def _resolve(cls, value: Any, context: dict[str, Any]) -> Any:
        if isinstance(value, dict) and set(value) == {"$ref"}:
            reference = value["$ref"]
            if not isinstance(reference, str):
                raise ValidationError("workflow $ref must be a string")
            return cls._lookup(context, reference)
        if isinstance(value, dict):
            return {key: cls._resolve(item, context) for key, item in value.items()}
        if isinstance(value, list):
            return [cls._resolve(item, context) for item in value]
        return value

    @staticmethod
    def _lookup(context: dict[str, Any], reference: str) -> Any:
        current: Any = context
        for part in reference.split("."):
            if not isinstance(current, dict) or part not in current:
                raise ValidationError(f"workflow reference not found: {reference}")
            current = current[part]
        return current

File: src/workflows.ts
import { randomUUID } from "node:crypto";
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { redact } from "./audit.js";
import { ValidationError } from "./errors.js";
import type { JsonObject, JsonValue } from "./models.js";
import { isRecord, utcNow } from "./models.js";
import { SkillRuntime } from "./skills.js";
import { Workspace } from "./storage.js";

export interface WorkflowStep { id: string; skill: string; input: JsonObject }
export interface WorkflowDefinition { id: string; version: string; name: string; steps: WorkflowStep[] }
export interface WorkflowRun { id: string; workflowId: string; workflowVersion: string; workspaceId?: string; status: "succeeded" | "failed"; startedAt: string; completedAt: string; input: JsonObject; steps: JsonObject[]; output: JsonObject; changeId?: string; error?: string }

export function loadWorkflow(path: string): WorkflowDefinition {
  return parseWorkflowDefinition(JSON.parse(readFileSync(path, "utf8")) as unknown);
}

export function parseWorkflowDefinition(value: unknown): WorkflowDefinition {
  if (!isRecord(value) || typeof value.id !== "string" || typeof value.version !== "string" || typeof value.name !== "string" || !Array.isArray(value.steps) || !value.steps.length) throw new ValidationError("workflow requires id, version, name, and non-empty steps");
  const seen = new Set<string>(); const steps = value.steps.map((raw): WorkflowStep => {
    if (!isRecord(raw) || typeof raw.id !== "string" || typeof raw.skill !== "string" || !isRecord(raw.input)) throw new ValidationError("each workflow step requires id, skill, and input");
    if (seen.has(raw.id)) throw new ValidationError(`duplicate workflow step id: ${raw.id}`); seen.add(raw.id); return { id: raw.id, skill: raw.skill, input: raw.input as JsonObject };
  });
  return { id: value.id, version: value.version, name: value.name, steps };
}

export class WorkflowRuntime {
  constructor(readonly workspace: Workspace, readonly skills: SkillRuntime) {}
  async run(definition: WorkflowDefinition, payload: JsonObject, options: { changeId?: string; workspaceId?: string; approvedSteps?: Set<string> } = {}): Promise<WorkflowRun> {
    const id = `workflow-run-${randomUUID().replaceAll("-", "")}`; const startedAt = utcNow(); const context: JsonObject = { input: payload, steps: {} }; const records: JsonObject[] = [];
    try {
      for (const step of definition.steps) {
        const resolved = this.resolve(step.input, context) as JsonObject; const run = await this.skills.run(step.skill, resolved, { ...(options.changeId ? { changeId: options.changeId } : {}), ...(options.workspaceId ? { workspaceId: options.workspaceId } : {}), approved: options.approvedSteps?.has(step.id) ?? false });
        (context.steps as JsonObject)[step.id] = { output: run.output, runId: run.id }; records.push({ id: step.id, skill: step.skill, skillRunId: run.id, status: run.status, output: run.output });
      }
      const last = definition.steps.at(-1); if (!last) throw new ValidationError("workflow has no steps");
      const run: WorkflowRun = { id, workflowId: definition.id, workflowVersion: definition.version, ...(options.changeId ? { changeId: options.changeId } : {}), ...(options.workspaceId ? { workspaceId: options.workspaceId } : {}), status: "succeeded", startedAt, completedAt: utcNow(), input: redact(payload) as JsonObject, steps: redact(records) as JsonObject[], output: redact(((context.steps as JsonObject)[last.id] as JsonObject).output ?? {}) as JsonObject };
      this.workspace.saveRun(run); return run;
    } catch (error) {
      const run: WorkflowRun = { id, workflowId: definition.id, workflowVersion: definition.version, ...(options.changeId ? { changeId: options.changeId } : {}), ...(options.workspaceId ? { workspaceId: options.workspaceId } : {}), status: "failed", startedAt, completedAt: utcNow(), input: redact(payload) as JsonObject, steps: redact(records) as JsonObject[], output: {}, error: `${error instanceof Error ? error.name : "Error"}: ${error instanceof Error ? error.message : String(error)}` };
      this.workspace.saveRun(run); throw error;
    }
  }
  private resolve(value: JsonValue, context: JsonObject): JsonValue {
    if (isRecord(value) && Object.keys(value).length === 1 && "$ref" in value) { if (typeof value.$ref !== "string") throw new ValidationError("workflow $ref must be a string"); return this.lookup(context, value.$ref); }
    if (Array.isArray(value)) return value.map((item) => this.resolve(item, context));
    if (isRecord(value)) return Object.fromEntries(Object.entries(value).map(([key, item]) => [key, this.resolve(item as JsonValue, context)]));
    return value;
  }
  private lookup(context: JsonObject, reference: string): JsonValue { let current: JsonValue = context; for (const part of reference.split(".")) { if (!isRecord(current) || !(part in current)) throw new ValidationError(`workflow reference not found: ${reference}`); current = current[part] as JsonValue; } return current; }
}
